### PR TITLE
keep resids of proteins if different or chains different

### PR DESCRIPTION
**Pull request by fgrunewald**
This is a draft fix for issue #366 . Basically if the -protres flag is given the merge molecule will only compute a residue offset if any of the resids overlap and any of the chain identifiers overlap. In this way you just keep residue offsets in the same chain if there is no overlap and you also keep resids of different chains in the same molecule. This is the expected behavior for most PDBs and proteins.

Code wise this branch needs some improvements. But I think it makes sense to have a flag with merge molecule that dictates protein behaviour, because it is quite specific to proteins. Also I was thinking to API wise if it would make sense to maybe have generic criteria for how to merge molecules.

### CODE REVIEW

File: bin/martinize2 [hunk 1]
```diff
@@ -360,6 +360,10 @@ def entry():
                             default=[], action='append', type=lambda x: x.split(','),
                             help='Ignore residues with that name: e.g. '
                             '-ignore HOH,LIG (+)')
+    file_group.add_argument('-protres', dest='protres', default=False,
+                            action='store_true',
+                            help="Check if molecules have the different chains "
+                                 " or resids when merging and preserve them if they are")
```
Comment (pckroon) [inline bin/martinize2:366]: Help is missing an 'else'. I like the functionality, but I don't like the name of the flag. Can we think of something better?

File: vermouth/molecule.py [hunk 2]
```diff
@@ -650,7 +650,7 @@ def __getattr__(self, name):
         else:
             raise AttributeError('Unknown attribute "{}".'.format(name))
 
-    def merge_molecule(self, molecule):
+    def merge_molecule(self, molecule, offset_handling='simple'):
```
Comment (pckroon) [inline vermouth/molecule.py:653]: New argument missing in docstring, and the CLI option does store_true, but this string implies we'll support multiple strategies, which will cause a change in CLI API at a later point. I'd suggest changing the CLI option to take a string with a sane default

File: vermouth/molecule.py [hunk 3]
```diff
@@ -681,16 +681,51 @@ def merge_molecule(self, molecule):
...
+        # this works well for many molecules but especially not
+        # for proteins where for example gaps in resid are common
+        if self.nodes() and offset_handling == 'simple':
             # We assume that the last id is always the largest.
             last_node_idx = max(self)
             offset = last_node_idx
             residue_offset = self.nodes[last_node_idx].get('resid', 1)
             offset_charge_group = self.nodes[last_node_idx].get('charge_group', 1)
+        # protein is more advanced and checks the intersection of chain and resid
+        # between the two merging molecules. If chains are all different, resids
```
Comment (pckroon) [inline vermouth/molecule.py:694]: This is also a valid approach for DNA and arbitrary linear polymers, no? Anything for which a sane chain can be identified

File: vermouth/molecule.py [hunk 4]
```diff
@@ -681,16 +681,51 @@ def merge_molecule(self, molecule):
...
+        # are preserved. If the chains are all the same but the resids are different
+        # resids are preserved. If both resid and chain are the same for at least one
+        # node we fall back to the simple scheme.
+        elif self.nodes() and offset_handling == 'protein':
+            current_chains = nx.get_node_attributes(self, 'chain')
+            new_chains = nx.get_node_attributes(molecule, 'chain')
+            same_chains = current_chains.items() & new_chains.items()
+            # all chains are the same
```
Comment (pckroon) [inline vermouth/molecule.py:702]: /some/ chains are the same

File: vermouth/molecule.py [hunk 5]
```diff
@@ -681,16 +681,51 @@ def merge_molecule(self, molecule):
...
+            if same_chains:
+                current_resids = nx.get_node_attributes(self, 'resid')
+                new_resids = nx.get_node_attributes(molecule, 'resid')
+                same_resids = current_resids.items() & new_resids.items()
+                if same_resids:
+                    last_node_idx = max(self)
```
Comment (pckroon) [inline vermouth/molecule.py:708]: Add comment here on what this if branch will do (and why)

File: vermouth/molecule.py [hunk 6]
```diff
@@ -681,16 +681,51 @@ def merge_molecule(self, molecule):
...
+                    offset = last_node_idx
+                    residue_offset = self.nodes[last_node_idx].get('resid', 1)
+                    offset_charge_group = self.nodes[last_node_idx].get('charge_group', 1)
+                else:
+                    last_node_idx = max(self)
```
Comment (pckroon) [inline vermouth/molecule.py:713]: Similar comment here

File: vermouth/molecule.py [hunk 7]
```diff
@@ -681,16 +681,51 @@ def merge_molecule(self, molecule):
...
+                    offset = last_node_idx
+                    residue_offset = 0
+                    # we keep this because GMX might use it and then they shouldn't be equal
+                    offset_charge_group = self.nodes[last_node_idx].get('charge_group', 1)
+            else:
```
Comment (pckroon) [inline vermouth/molecule.py:718]: And also here

File: vermouth/molecule.py (hunk 5 above)
Comment (pckroon) [inline vermouth/molecule.py:708]: You need a last_node_idx for all cases, so move it up

File: vermouth/processors/merge_chains.py [hunk 8]
```diff
@@ -22,7 +22,7 @@
 from ..processors.processor import Processor
 
 
-def merge_chains(system, chains):
+def merge_chains(system, chains, offset_handling='simple'):
```
Comment (pckroon) [inline vermouth/processors/merge_chains.py:25]: Missing in docstring

**Review by pckroon — CHANGES_REQUESTED** [review]
Overall looks quite good, but I have a few minor concerns on design. Also, tests and docstrings are missing (but this is still a draft, so fine for now).
The bigger discussion is what attributes we want to use to identify a residue (monomeric repeat unit). Chain + resname + resid sounds reasonable, but then we miss the case where there are no chain IDs, but there is a TER record between two chains. We can consider "fixing" the chain IDs in cases like that, or add a molecule_idx as identifier.

**Comment by fgrunewald** [comment]
a better PR is on the way

**Pull request closed without merging**